# Тестовое задание на Junior ML Engineer в команду Матчинга в Профи (Вариант 2)
## Абаполов Филипп

## Задание 1
__Условие:__

Подкинули монету N раз. Кол-во случаев, когда выпал орёл, на 10% больше, чем кол-во случаев, когда выпала решка. При каком N мы можем сказать, что монета «нечестная» (орёл и решка выпадают с разной вероятностью)?

__Решение__

$x$ - количество выпаданий решки, тогда $1.1*x$ - количество выпаданий орла.
$x+1.1x=2.1x=N \Rightarrow N$ кратно 21.

Возьмем уровень значимости $\alpha = 0.05$, нулевая гипотеза $H_0$ - монетка честная.
С помощью Python я перебираю различные N, кратные 21 и считаю p-value, используя binom\_test из scipy.stats. И вывожу минимальное N, при котором p-value меньше критического значения $\alpha$

In [12]:
from scipy.stats import binom_test
alpha = 0.05
N = 21

while True:
    p_value = binom_test(N/2.1, N, p = 0.5)
    if p_value < alpha:
        print(f'N = {N}')
        break
    N += 21

N = 1743


При $N=1743$ мы можем отвергнуть нулевую гипотезу $H_0$.

Ответ:

$N=1743$

## Задание 2
__Условие:__

Есть две монеты: «правильная» (с одной стороны орёл, с другой решка) и «неправильная» (орёл на обеих сторонах). Берём одну монету наугад и подкидываем 5 раз. Все 5 раз выпадает орёл. Какова вероятность того, что выбрали «правильную» монету?

__Решение:__

$A$ - взяли правильную монету, $P(A)=\frac{2}{3}$, $\overline{A}$ - взяли неправильную монету, $P(\overline{A})=\frac{1}{3}$.

$X$ - выпало 5 орлов подряд.

Тогда $P(X|A)=0.5^5$, $P(X|\overline{A})=1$

По формуле полной вероятности:
$$P(X)=P(X|A)P(A)+P(X|\overline{A})P(\overline{A})=\frac{2}{3}0.5^5+\frac{1}{3}1=\frac{17}{48}$$

Тогда:

$$P(A|X)=\frac{P(X|A)*P(A)}{P(X)}=\frac{0.5^5\frac{2}{3}}{\frac{17}{48}}=\frac{1}{17}\approx 0.0588$$

Ответ:

$P(A|X)\approx 0.0588$